File: src/rules-compiler-rust/src/compiler.rs
//! Core compiler functionality for AdGuard filter rules.
//!
//! This module provides the main compilation logic, wrapping the hostlist-compiler
//! tool and providing statistics, hashing, and file management.

use chrono::{DateTime, Utc};
use sha2::{Digest, Sha384};
use std::fs::{self, File};
use std::io::{BufRead, BufReader, Read};
use std::path::{Path, PathBuf};
use std::process::Command;
use std::time::Instant;

use crate::config::{read_config, to_json, CompilerConfig, ConfigFormat};
use crate::error::{CompilerError, Result};

/// Platform-specific information.
#[derive(Debug, Clone, Default)]
pub struct PlatformInfo {
    /// Operating system name.
    pub os_name: String,
    /// Operating system version.
    pub os_version: String,
    /// Processor architecture.
    pub architecture: String,
    /// Whether the platform is Windows.
    pub is_windows: bool,
    /// Whether the platform is Linux.
    pub is_linux: bool,
    /// Whether the platform is macOS.
    pub is_macos: bool,
}

impl PlatformInfo {
    /// Detect current platform information.
    #[must_use]
    pub fn detect() -> Self {
        Self {
            os_name: std::env::consts::OS.to_string(),
            os_version: String::new(),
            architecture: std::env::consts::ARCH.to_string(),
            is_windows: cfg!(target_os = "windows"),
            is_linux: cfg!(target_os = "linux"),
            is_macos: cfg!(target_os = "macos"),
        }
    }
}

/// Version information for all components.
#[derive(Debug, Clone, Default)]
pub struct VersionInfo {
    /// Module version.
    pub module_version: String,
    /// Rust version.
    pub rust_version: String,
    /// Node.js version (if available).
    pub node_version: Option<String>,
    /// hostlist-compiler version (if available).
    pub hostlist_compiler_version: Option<String>,
    /// Path to hostlist-compiler.
    pub hostlist_compiler_path: Option<String>,
    /// Platform information.
    pub platform: PlatformInfo,
}

impl VersionInfo {
    /// Collect version information for all components.
    #[must_use]
    pub fn collect() -> Self {
        let mut info = Self {
            module_version: crate::VERSION.to_string(),
            rust_version: format!("{}", rustc_version_runtime::version()),
            platform: PlatformInfo::detect(),
            ..Default::default()
        };

        // Check Node.js
        if let Some(node_path) = find_command("node") {
            info.node_version =
                get_command_version(node_path.to_str().unwrap_or("node"), &["--version"]);
        }

        // Check hostlist-compiler
        if let Some(compiler_path) = find_command("hostlist-compiler") {
            info.hostlist_compiler_path = Some(compiler_path.display().to_string());
            info.hostlist_compiler_version = get_command_version(
                compiler_path.to_str().unwrap_or("hostlist-compiler"),
                &["--version"],
            );
        } else if find_command("npx").is_some() {
            info.hostlist_compiler_path = Some("npx @adguard/hostlist-compiler".to_string());
        }

        info
    }

    /// Check if hostlist-compiler is available.
    #[must_use]
    pub fn has_compiler(&self) -> bool {
        self.hostlist_compiler_path.is_some()
    }

    /// Check if Node.js is available.
    #[must_use]
    pub fn has_node(&self) -> bool {
        self.node_version.is_some()
    }
}

/// Result of a compilation operation.
#[derive(Debug, Clone)]
pub struct CompilerResult {
    /// Whether compilation was successful.
    pub success: bool,
    /// Name from configuration.
    pub config_name: String,
    /// Version from configuration.
    pub config_version: String,
    /// Number of rules in output.
    pub rule_count: usize,
    /// Path to output file.
    pub output_path: PathBuf,
    /// SHA-384 hash of output file.
    pub output_hash: String,
    /// Whether output was copied to rules directory.
    pub copied_to_rules: bool,
    /// Destination path if copied.
    pub rules_destination: Option<PathBuf>,
    /// Elapsed time in milliseconds.
    pub elapsed_ms: u64,
    /// Start time.
    pub start_time: DateTime<Utc>,
    /// End time.
    pub end_time: DateTime<Utc>,
    /// Error message if failed.
    pub error_message: Option<String>,
    /// Standard output from compiler.
    pub stdout: String,
    /// Standard error from compiler.
    pub stderr: String,
}

impl Default for CompilerResult {
    fn default() -> Self {
        let now = Utc::now();
        Self {
            success: false,
            config_name: String::new(),
            config_version: String::new(),
            rule_count: 0,
            output_path: PathBuf::new(),
            output_hash: String::new(),
            copied_to_rules: false,
            rules_destination: None,
            elapsed_ms: 0,
            start_time: now,
            end_time: now,
            error_message: None,
            stdout: String::new(),
            stderr: String::new(),
        }
    }
}

impl CompilerResult {
    /// Get the output path as a string.
    #[must_use]
    pub fn output_path_str(&self) -> String {
        self.output_path.display().to_string()
    }

    /// Get the rules destination path as a string.
    #[must_use]
    pub fn rules_destination_str(&self) -> Option<String> {
        self.rules_destination
            .as_ref()
            .map(|p| p.display().to_string())
    }

    /// Get elapsed time as a formatted string.
    #[must_use]
    pub fn elapsed_formatted(&self) -> String {
        if self.elapsed_ms >= 1000 {
            format!("{:.2}s", self.elapsed_ms as f64 / 1000.0)
        } else {
            format!("{}ms", self.elapsed_ms)
        }
    }

    /// Get truncated hash for display.
    #[must_use]
    pub fn hash_short(&self) -> &str {
        if self.output_hash.len() >= 32 {
            &self.output_hash[..32]
        } else {
            &self.output_hash
        }
    }
}

/// Options for running the compiler.
#[derive(Debug, Clone, Default)]
pub struct CompileOptions {
    /// Path to output file (auto-generated if None).
    pub output_path: Option<PathBuf>,
    /// Copy output to rules directory.
    pub copy_to_rules: bool,
    /// Custom rules directory.
    pub rules_directory: Option<PathBuf>,
    /// Force configuration format.
    pub format: Option<ConfigFormat>,
    /// Enable debug output.
    pub debug: bool,
    /// Validate configuration before compiling.
    pub validate: bool,
    /// Fail compilation on validation warnings.
    pub fail_on_warnings: bool,
}

impl CompileOptions {
    /// Create new compile options with default values.
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    /// Set the output path.
    #[must_use]
    pub fn with_output<P: Into<PathBuf>>(mut self, path: P) -> Self {
        self.output_path = Some(path.into());
        self
    }

    /// Enable copying to rules directory.
    #[must_use]
    pub const fn with_copy_to_rules(mut self, copy: bool) -> Self {
        self.copy_to_rules = copy;
        self
    }

    /// Set the rules directory.
    #[must_use]
    pub fn with_rules_directory<P: Into<PathBuf>>(mut self, path: P) -> Self {
        self.rules_directory = Some(path.into());
        self
    }

    /// Set the configuration format.
    #[must_use]
    pub const fn with_format(mut self, format: ConfigFormat) -> Self {
        self.format = Some(format);
        self
    }

    /// Enable debug output.
    #[must_use]
    pub const fn with_debug(mut self, debug: bool) -> Self {
        self.debug = debug;
        self
    }

    /// Enable validation.
    #[must_use]
    pub const fn with_validation(mut self, validate: bool) -> Self {
        self.validate = validate;
        self
    }

    /// Set fail on warnings.
    #[must_use]
    pub const fn with_fail_on_warnings(mut self, fail_on_warnings: bool) -> Self {
        self.fail_on_warnings = fail_on_warnings;
        self
    }
}

/// Main compiler for AdGuard filter rules.
#[derive(Debug, Default)]
pub struct RulesCompiler {
    options: CompileOptions,
}

impl RulesCompiler {
    /// Create a new compiler instance with default options.
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    /// Create a new compiler instance with custom options.
    #[must_use]
    pub const fn with_options(options: CompileOptions) -> Self {
        Self { options }
    }

    /// Get mutable reference to options.
    pub fn options_mut(&mut self) -> &mut CompileOptions {
        &mut self.options
    }

    /// Compile filter rules from a configuration file.
    ///
    /// # Errors
    ///
    /// Returns an error if compilation fails.
    pub fn compile<P: AsRef<Path>>(&self, config_path: P) -> Result<CompilerResult> {
        compile_rules(config_path, &self.options)
    }

    /// Read configuration from a file.
    ///
    /// # Errors
    ///
    /// Returns an error if the file can't be read or parsed.
    pub fn read_config<P: AsRef<Path>>(&self, config_path: P) -> Result<CompilerConfig> {
        read_config(config_path, self.options.format)
    }

    /// Get version information.
    #[must_use]
    pub fn version_info(&self) -> VersionInfo {
        VersionInfo::collect()
    }
}

/// Find command in PATH.
fn find_command(name: &str) -> Option<PathBuf> {
    which::which(name).ok()
}

/// Get version from a command.
fn get_command_version(cmd: &str, args: &[&str]) -> Option<String> {
    Command::new(cmd)
        .args(args)
        .output()
        .ok()
        .and_then(|output| {
            if output.status.success() {
                String::from_utf8(output.stdout)
                    .ok()
                    .map(|s| s.lines().next().unwrap_or("").trim().to_string())
            } else {
                None
            }
        })
}

/// Count non-empty, non-comment lines in a file.
///
/// Lines starting with `!` or `#` are considered comments.
#[must_use]
pub fn count_rules<P: AsRef<Path>>(path: P) -> usize {
    let file = match File::open(path.as_ref()) {
        Ok(f) => f,
        Err(_) => return 0,
    };

    BufReader::new(file)
        .lines()
        .map_while(std::result::Result::ok)
        .filter(|line| {
            let trimmed = line.trim();
            !trimmed.is_empty() && !trimmed.starts_with('!') && !trimmed.starts_with('#')
        })
        .count()
}

/// Compute SHA-384 hash of a file.
///
/// # Errors
///
/// Returns an error if the file can't be read.
pub fn compute_hash<P: AsRef<Path>>(path: P) -> Result<String> {
    let mut file = File::open(path.as_ref())?;
    let mut hasher = Sha384::new();
    let mut buffer = [0u8; 8192];

    loop {
        let bytes_read = file.read(&mut buffer)?;
        if bytes_read == 0 {
            break;
        }
        hasher.update(&buffer[..bytes_read]);
    }

    Ok(hex::encode(hasher.finalize()))
}

/// Get compiler command and arguments.
fn get_compiler_command(config_path: &str, output_path: &str) -> Result<(String, Vec<String>)> {
    if let Some(compiler_path) = find_command("hostlist-compiler") {
        return Ok((
            compiler_path.display().to_string(),
            vec![
                "--config".to_string(),
                config_path.to_string(),
                "--output".to_string(),
                output_path.to_string(),
            ],
        ));
    }

    if let Some(npx_path) = find_command("npx") {
        return Ok((
            npx_path.display().to_string(),
            vec![
                "@adguard/hostlist-compiler".to_string(),
                "--config".to_string(),
                config_path.to_string(),
                "--output".to_string(),
                output_path.to_string(),
            ],
        ));
    }

    Err(CompilerError::CompilerNotFound)
}

/// Generate default output path based on config path and timestamp.
fn generate_output_path(config_path: &Path) -> PathBuf {
    let timestamp = Utc::now().format("%Y%m%d-%H%M%S");
    let output_dir = config_path
        .parent()
        .unwrap_or(Path::new("."))
        .join("output");
    output_dir.join(format!("compiled-{timestamp}.txt"))
}

/// Determine rules directory from config path.
fn get_rules_directory(config_path: &Path, custom: Option<&Path>) -> PathBuf {
    custom.map(Path::to_path_buf).unwrap_or_else(|| {
        config_path
            .parent()
            .unwrap_or(Path::new("."))
            .parent()
            .unwrap_or(Path::new("."))
            .parent()
            .unwrap_or(Path::new("."))
            .join("rules")
    })
}

/// Compile filter rules using hostlist-compiler.
///
/// # Arguments
///
/// * `config_path` - Path to the configuration file.
/// * `options` - Compilation options.
///
/// # Errors
///
/// Returns an error if compilation fails.
pub fn compile_rules<P: AsRef<Path>>(
    config_path: P,
    options: &CompileOptions,
) -> Result<CompilerResult> {
    let start = Instant::now();
    let mut result = CompilerResult {
        start_time: Utc::now(),
        ..Default::default()
    };

    let config_path = config_path.as_ref().canonicalize().map_err(|e| {
        CompilerError::file_system(
            format!("resolving config path {}", config_path.as_ref().display()),
            e,
        )
    })?;

    // Read configuration
    let config = read_config(&config_path, options.format)?;
    result.config_name = config.name.clone();
    result.config_version = config.version.clone();

    // Validate if requested
    if options.validate {
        config.validate()?;
    }

    // Determine output path
    let output_path = options
        .output_path
        .clone()
        .unwrap_or_else(|| generate_output_path(&config_path));
    result.output_path = output_path.clone();

    // Convert to JSON if needed (hostlist-compiler only accepts JSON)
    let (compile_config_path, temp_config_path) = if config.format() != Some(ConfigFormat::Json) {
        let temp_path =
            std::env::temp_dir().join(format!("compiler-config-{}.json", uuid::Uuid::new_v4()));
        let json = to_json(&config)?;
        fs::write(&temp_path, &json).map_err(|e| {
            CompilerError::file_system(format!("writing temp config to {}", temp_path.display()), e)
        })?;

        if options.debug {
            eprintln!("[DEBUG] Created temp JSON config: {}", temp_path.display());
            eprintln!("[DEBUG] Config content:\n{json}");
        }

        (temp_path.clone(), Some(temp_path))
    } else {
        (config_path.clone(), None)
    };

    // Ensure output directory exists
    if let Some(output_dir) = output_path.parent() {
        fs::create_dir_all(output_dir).map_err(|e| {
            CompilerError::file_system(
                format!("creating output directory {}", output_dir.display()),
                e,
            )
        })?;
    }

    // Get compiler command
    let (cmd, args) = get_compiler_command(
        compile_config_path.to_str().unwrap_or(""),
        output_path.to_str().unwrap_or(""),
    )?;

    if options.debug {
        eprintln!("[DEBUG] Running: {cmd} {}", args.join(" "));
    }

    // Run compilation
    let output = Command::new(&cmd)
        .args(&args)
        .current_dir(config_path.parent().unwrap_or(Path::new(".")))
        .output()
        .map_err(|e| CompilerError::process_execution(format!("{cmd} {}", args.join(" ")), e))?;

    result.stdout = String::from_utf8_lossy(&output.stdout).to_string();
    result.stderr = String::from_utf8_lossy(&output.stderr).to_string();

    // Clean up temp file
    if let Some(temp_path) = temp_config_path {
        let _ = fs::remove_file(temp_path);
    }

    // Check for compilation failure
    if !output.status.success() {
        result.error_message = Some(format!(
            "compiler exited with code {:?}: {}",
            output.status.code(),
            result.stderr.trim()
        ));
        result.end_time = Utc::now();
        result.elapsed_ms = start.elapsed().as_millis() as u64;
        return Ok(result);
    }

    // Verify output was created
    if !output_path.exists() {
        result.error_message = Some("output file was not created".to_string());
        result.end_time = Utc::now();
        result.elapsed_ms = start.elapsed().as_millis() as u64;
        return Ok(result);
    }

    // Calculate statistics
    result.rule_count = count_rules(&output_path);
    result.output_hash = compute_hash(&output_path)?;
    result.success = true;

    // Copy to rules directory if requested
    if options.copy_to_rules {
        let rules_dir = get_rules_directory(&config_path, options.rules_directory.as_deref());
        fs::create_dir_all(&rules_dir).map_err(|e| {
            CompilerError::file_system(
                format!("creating rules directory {}", rules_dir.display()),
                e,
            )
        })?;

        let dest_path = rules_dir.join("adguard_user_filter.txt");
        fs::copy(&output_path, &dest_path).map_err(|e| {
            CompilerError::copy_failed(
                format!(
                    "copying {} to {}",
                    output_path.display(),
                    dest_path.display()
                ),
                e,
            )
        })?;

        result.copied_to_rules = true;
        result.rules_destination = Some(dest_path);
    }

    result.end_time = Utc::now();
    result.elapsed_ms = start.elapsed().as_millis() as u64;

    Ok(result)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;
    use tempfile::TempDir;

    #[test]
    fn test_platform_info_detect() {
        let info = PlatformInfo::detect();
        assert!(!info.os_name.is_empty());
        assert!(!info.architecture.is_empty());
    }

    #[test]
    fn test_version_info_collect() {
        let info = VersionInfo::collect();
        assert!(!info.module_version.is_empty());
        assert!(!info.rust_version.is_empty());
    }

    #[test]
    fn test_count_rules() {
        let dir = TempDir::new().unwrap();
        let path = dir.path().join("rules.txt");
        let mut file = File::create(&path).unwrap();
        writeln!(file, "! Comment").unwrap();
        writeln!(file, "# Another comment").unwrap();
        writeln!(file, "||example.com^").unwrap();
        writeln!(file, "||test.org^").unwrap();
        writeln!(file).unwrap();
        writeln!(file, "@@||allowed.com^").unwrap();

        assert_eq!(count_rules(&path), 3);
    }

    #[test]
    fn test_count_rules_empty_file() {
        let dir = TempDir::new().unwrap();
        let path = dir.path().join("empty.txt");
        File::create(&path).unwrap();

        assert_eq!(count_rules(&path), 0);
    }

    #[test]
    fn test_count_rules_nonexistent() {
        assert_eq!(count_rules("/nonexistent/path.txt"), 0);
    }

    #[test]
    fn test_compute_hash() {
        let dir = TempDir::new().unwrap();
        let path = dir.path().join("test.txt");
        let mut file = File::create(&path).unwrap();
        writeln!(file, "Test content").unwrap();

        let hash = compute_hash(&path).unwrap();
        assert_eq!(hash.len(), 96); // SHA-384 = 96 hex chars
    }

    #[test]
    fn test_compile_options_builder() {
        let options = CompileOptions::new()
            .with_output("/output/path.txt")
            .with_copy_to_rules(true)
            .with_debug(true)
            .with_validation(true);

        assert_eq!(options.output_path, Some(PathBuf::from("/output/path.txt")));
        assert!(options.copy_to_rules);
        assert!(options.debug);
        assert!(options.validate);
    }

    #[test]
    fn test_compiler_result_helpers() {
        let mut result = CompilerResult::default();
        result.output_path = PathBuf::from("/path/to/output.txt");
        result.output_hash = "a".repeat(96);
        result.elapsed_ms = 1500;

        assert_eq!(result.output_path_str(), "/path/to/output.txt");
        assert_eq!(result.hash_short().len(), 32);
        assert_eq!(result.elapsed_formatted(), "1.50s");

        result.elapsed_ms = 500;
        assert_eq!(result.elapsed_formatted(), "500ms");
    }

    #[test]
    fn test_generate_output_path() {
        let config_path = PathBuf::from("/project/config/compiler.json");
        let output_path = generate_output_path(&config_path);
        assert!(output_path.to_str().unwrap().contains("compiled-"));
        assert!(output_path.to_str().unwrap().ends_with(".txt"));
    }
}

File: src/rules-compiler-python/rules_compiler/compiler.py
"""
Core compiler functionality for AdGuard filter rules.
"""

from __future__ import annotations

import hashlib
import json
import logging
import os
import platform
import shutil
import subprocess
import tempfile
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Any

from rules_compiler.config import (
    CompilerConfiguration,
    ConfigurationFormat,
    read_configuration,
    to_json,
)
from rules_compiler.errors import (
    CompilationError,
    CompilerNotFoundError,
    CopyError,
    OutputNotCreatedError,
    TimeoutError as CompilerTimeoutError,
    ValidationError,
)

logger = logging.getLogger(__name__)


@dataclass
class PlatformInfo:
    """Platform-specific information."""
    os_name: str = ""
    os_version: str = ""
    architecture: str = ""
    is_windows: bool = False
    is_linux: bool = False
    is_macos: bool = False


@dataclass
class VersionInfo:
    """Version information for all components."""
    module_version: str = ""
    python_version: str = ""
    node_version: str | None = None
    hostlist_compiler_version: str | None = None
    hostlist_compiler_path: str | None = None
    platform: PlatformInfo = field(default_factory=PlatformInfo)

    def has_node(self) -> bool:
        """Check if Node.js is available."""
        return self.node_version is not None

    def has_compiler(self) -> bool:
        """Check if hostlist-compiler is available."""
        return self.hostlist_compiler_version is not None

    @classmethod
    def collect(cls) -> VersionInfo:
        """Collect version information from the system."""
        return get_version_info()


@dataclass
class CompilerResult:
    """Result of a compilation operation."""
    success: bool = False
    config_name: str = ""
    config_version: str = ""
    rule_count: int = 0
    output_path: str = ""
    output_hash: str = ""
    copied_to_rules: bool = False
    rules_destination: str | None = None
    elapsed_ms: int = 0
    start_time: datetime = field(default_factory=datetime.utcnow)
    end_time: datetime = field(default_factory=datetime.utcnow)
    error_message: str | None = None
    stdout: str = ""
    stderr: str = ""

    def elapsed_formatted(self) -> str:
        """Get elapsed time in human-readable format."""
        if self.elapsed_ms >= 1000:
            return f"{self.elapsed_ms / 1000:.2f}s"
        return f"{self.elapsed_ms}ms"

    def hash_short(self, length: int = 32) -> str:
        """Get shortened hash for display."""
        if len(self.output_hash) > length:
            return self.output_hash[:length]
        return self.output_hash

    def output_path_str(self) -> str:
        """Get output path as string."""
        return str(self.output_path) if self.output_path else ""

    def rules_destination_str(self) -> str | None:
        """Get rules destination as string or None."""
        return str(self.rules_destination) if self.rules_destination else None


def get_platform_info() -> PlatformInfo:
    """Get information about the current platform."""
    system = platform.system().lower()
    return PlatformInfo(
        os_name=platform.system(),
        os_version=platform.version(),
        architecture=platform.machine(),
        is_windows=system == "windows",
        is_linux=system == "linux",
        is_macos=system == "darwin",
    )


def find_command(command: str) -> str | None:
    """Find a command in PATH."""
    return shutil.which(command)


def get_version_info() -> VersionInfo:
    """Get version information for all components."""
    from rules_compiler import __version__

    info = VersionInfo(
        module_version=__version__,
        python_version=platform.python_version(),
        platform=get_platform_info(),
    )

    # Check Node.js
    node_path = find_command("node")
    if node_path:
        try:
            result = subprocess.run(
                [node_path, "--version"],
                capture_output=True,
                text=True,
                timeout=10,
            )
            if result.returncode == 0:
                info.node_version = result.stdout.strip()
        except Exception:
            pass

    # Check hostlist-compiler
    compiler_path = find_command("hostlist-compiler")
    if compiler_path:
        info.hostlist_compiler_path = compiler_path
        try:
            result = subprocess.run(
                [compiler_path, "--version"],
                capture_output=True,
                text=True,
                timeout=10,
            )
            if result.returncode == 0:
                info.hostlist_compiler_version = result.stdout.strip().split("\n")[0]
        except Exception:
            pass
    else:
        # Check npx
        npx_path = find_command("npx")
        if npx_path:
            info.hostlist_compiler_path = f"{npx_path} @adguard/hostlist-compiler"

    return info


def count_rules(file_path: str | Path) -> int:
    """
    Count non-empty, non-comment lines in a file.

    Args:
        file_path: Path to the file.

    Returns:
        Number of rules.
    """
    path = Path(file_path)
    if not path.exists():
        return 0

    count = 0
    with open(path, "r", encoding="utf-8") as f:
        for line in f:
            stripped = line.strip()
            if stripped and not stripped.startswith(("!", "#")):
                count += 1

    return count


def compute_hash(file_path: str | Path) -> str:
    """
    Compute SHA-384 hash of a file.

    Args:
        file_path: Path to the file.

    Returns:
        Hex-encoded hash string.
    """
    sha384 = hashlib.sha384()
    with open(file_path, "rb") as f:
        for chunk in iter(lambda: f.read(8192), b""):
            sha384.update(chunk)
    return sha384.hexdigest()


def hash_short(hash_value: str, length: int = 32) -> str:
    """
    Get shortened hash for display.

    Args:
        hash_value: Full hash string.
        length: Maximum length to return.

    Returns:
        Shortened hash string.
    """
    if len(hash_value) > length:
        return hash_value[:length]
    return hash_value


def format_elapsed(elapsed_ms: int) -> str:
    """
    Format elapsed time in human-readable format.

    Args:
        elapsed_ms: Elapsed time in milliseconds.

    Returns:
        Formatted string (e.g., "1.50s" or "500ms").
    """
    if elapsed_ms >= 1000:
        return f"{elapsed_ms / 1000:.2f}s"
    return f"{elapsed_ms}ms"


def _get_compiler_command(config_path: str, output_path: str) -> tuple[list[str], str]:
    """
    Get the compiler command and working directory.

    Returns:
        Tuple of (command args, working directory).

    Raises:
        CompilerNotFoundError: If hostlist-compiler is not found.
    """
    compiler_path = find_command("hostlist-compiler")

    if compiler_path:
        return (
            [compiler_path, "--config", config_path, "--output", output_path],
            str(Path(config_path).parent),
        )

    npx_path = find_command("npx")
    if npx_path:
        return (
            [npx_path, "@adguard/hostlist-compiler", "--config", config_path, "--output", output_path],
            str(Path(config_path).parent),
        )

    raise CompilerNotFoundError(["hostlist-compiler", "npx"])


class RulesCompiler:
    """
    Main compiler class for AdGuard filter rules.

    Example:
        >>> compiler = RulesCompiler()
        >>> result = compiler.compile("config.yaml", copy_to_rules=True)
        >>> if result.success:
        ...     print(f"Compiled {result.rule_count} rules")
    """

    def __init__(self, debug: bool = False):
        """
        Initialize the compiler.

        Args:
            debug: Enable debug logging.
        """
        self.debug = debug
        if debug:
            logging.basicConfig(level=logging.DEBUG)

    def compile(
        self,
        config_path: str | Path,
        output_path: str | Path | None = None,
        copy_to_rules: bool = False,
        rules_directory: str | Path | None = None,
        format: ConfigurationFormat | None = None,
        validate: bool = True,
    ) -> CompilerResult:
        """
        Compile filter rules.

        Args:
            config_path: Path to configuration file.
            output_path: Optional output file path.
            copy_to_rules: Copy output to rules directory.
            rules_directory: Custom rules directory path.
            format: Force configuration format.
            validate: Validate configuration before compiling.

        Returns:
            Compilation result.
        """
        return compile_rules(
            config_path=config_path,
            output_path=output_path,
            copy_to_rules=copy_to_rules,
            rules_directory=rules_directory,
            format=format,
            debug=self.debug,
            validate=validate,
        )

    def read_config(
        self,
        config_path: str | Path,
        format: ConfigurationFormat | None = None,
    ) -> CompilerConfiguration:
        """
        Read configuration from a file.

        Args:
            config_path: Path to configuration file.
            format: Force configuration format.

        Returns:
            Parsed configuration.
        """
        return read_configuration(config_path, format)

    def validate_config(
        self,
        config: CompilerConfiguration,
        check_files: bool = False,
    ) -> tuple[bool, list[str], list[str]]:
        """
        Validate a configuration.

        Args:
            config: Configuration to validate.
            check_files: Check if local source files exist.

        Returns:
            Tuple of (is_valid, errors, warnings).
        """
        result = config.validate(check_files=check_files)
        return result.is_valid, result.errors, result.warnings

    def get_version_info(self) -> VersionInfo:
        """Get version information for all components."""
        return get_version_info()


def compile_rules(
    config_path: str | Path,
    output_path: str | Path | None = None,
    copy_to_rules: bool = False,
    rules_directory: str | Path | None = None,
    format: ConfigurationFormat | None = None,
    debug: bool = False,
    validate: bool = True,
) -> CompilerResult:
    """
    Compile filter rules using hostlist-compiler.

    Args:
        config_path: Path to configuration file.
        output_path: Optional output file path.
        copy_to_rules: Copy output to rules directory.
        rules_directory: Custom rules directory path.
        format: Force configuration format.
        debug: Enable debug logging.
        validate: Validate configuration before compiling.

    Returns:
        Compilation result.
    """
    result = CompilerResult(start_time=datetime.utcnow())
    config_path = Path(config_path).resolve()
    temp_config_path = None

    try:
        # Read configuration
        config = read_configuration(config_path, format)
        result.config_name = config.name
        result.config_version = config.version

        # Validate configuration if requested
        if validate:
            validation_result = config.validate()
            if not validation_result.is_valid:
                raise ValidationError(validation_result.errors, validation_result.warnings)
            if validation_result.warnings and debug:
                for warning in validation_result.warnings:
                    logger.warning(f"Config warning: {warning}")

        # Determine output path
        if output_path:
            actual_output = Path(output_path).resolve()
        else:
            timestamp = datetime.utcnow().strftime("%Y%m%d-%H%M%S")
            output_dir = config_path.parent / "output"
            output_dir.mkdir(exist_ok=True)
            actual_output = output_dir / f"compiled-{timestamp}.txt"

        result.output_path = str(actual_output)

        # Convert to JSON if needed (hostlist-compiler only supports JSON)
        detected_format = format or config._source_format
        if detected_format != ConfigurationFormat.JSON:
            temp_config_path = tempfile.NamedTemporaryFile(
                mode="w",
                suffix=".json",
                delete=False,
                encoding="utf-8",
            )
            temp_config_path.write(to_json(config))
            temp_config_path.close()
            compile_config_path = temp_config_path.name
            if debug:
                logger.debug(f"Created temp JSON config: {compile_config_path}")
        else:
            compile_config_path = str(config_path)

        # Get compiler command
        cmd, cwd = _get_compiler_command(compile_config_path, str(actual_output))

        if debug:
            logger.debug(f"Running: {' '.join(cmd)}")
            logger.debug(f"Working directory: {cwd}")

        # Run compilation
        try:
            proc = subprocess.run(
                cmd,
                cwd=cwd,
                capture_output=True,
                text=True,
                timeout=300,  # 5 minute timeout
            )
        except subprocess.TimeoutExpired:
            raise CompilerTimeoutError(300, " ".join(cmd))

        result.stdout = proc.stdout
        result.stderr = proc.stderr

        if proc.returncode != 0:
            raise CompilationError(
                f"Compiler exited with code {proc.returncode}",
                exit_code=proc.returncode,
                stdout=proc.stdout,
                stderr=proc.stderr,
            )

        # Verify output was created
        if not actual_output.exists():
            raise OutputNotCreatedError(str(actual_output))

        # Calculate statistics
        result.rule_count = count_rules(actual_output)
        result.output_hash = compute_hash(actual_output)
        result.success = True

        if debug:
            logger.debug(f"Compiled {result.rule_count} rules")
            logger.debug(f"Output hash: {result.hash_short()}...")

        # Copy to rules directory if requested
        if copy_to_rules:
            if rules_directory:
                rules_dir = Path(rules_directory)
            else:
                rules_dir = config_path.parent.parent.parent / "rules"

            try:
                rules_dir.mkdir(exist_ok=True)
                dest_path = rules_dir / "adguard_user_filter.txt"
                shutil.copy2(actual_output, dest_path)
                result.copied_to_rules = True
                result.rules_destination = str(dest_path)
                if debug:
                    logger.debug(f"Copied to: {dest_path}")
            except (OSError, IOError) as e:
                raise CopyError(str(actual_output), str(dest_path), str(e))

    except (ValidationError, CompilationError, CompilerNotFoundError,
            OutputNotCreatedError, CopyError, CompilerTimeoutError) as e:
        result.success = False
        result.error_message = str(e)
        logger.error(f"Compilation failed: {e}")

    except Exception as e:
        result.success = False
        result.error_message = str(e)
        logger.error(f"Compilation failed: {e}")

    finally:
        # Clean up temp file
        if temp_config_path and os.path.exists(temp_config_path.name):
            os.unlink(temp_config_path.name)

        result.end_time = datetime.utcnow()
        result.elapsed_ms = int((result.end_time - result.start_time).total_seconds() * 1000)

    return result


def validate_configuration(
    config_path: str | Path,
    format: ConfigurationFormat | None = None,
    check_files: bool = False,
) -> tuple[bool, list[str], list[str]]:
    """
    Validate a configuration file without compiling.

    Args:
        config_path: Path to configuration file.
        format: Force configuration format.
        check_files: Check if local source files exist.

    Returns:
        Tuple of (is_valid, errors, warnings).
    """
    config = read_configuration(config_path, format)
    result = config.validate(check_files=check_files)
    return result.is_valid, result.errors, result.warnings
